File: zig/glob.zig
const std = @import("std");
const print = std.debug.print;
const expect = std.testing.expect;
var alloc = std.heap.GeneralPurposeAllocator(.{}){};

fn glob(pattern: []const u8, text: []const u8) bool {
    var p: usize = 0;
    var t: usize = 0;
    var np: usize = 0;
    var nt: usize = 0;
    while (p < pattern.len or t < text.len) {
        if (p < pattern.len) {
            switch (pattern[p]) {
                '*' => {
                    np = p;
                    nt = t + 1;
                    p += 1;
                    continue;
                },
                '?' => {
                    if (nt < text.len) {
                        p += 1;
                        t += 1;
                        continue;
                    }
                },
                else => {
                    if (t < text.len and text[t] == pattern[p]) {
                        p += 1;
                        t += 1;
                        continue;
                    }
                },
            }
        }
        if (nt > 0 and nt <= text.len) {
            p = np;
            t = nt;
            continue;
        }
        return false;
    }
    return true;
}

test "glob" {
    expect(glob("", ""));
    expect(glob("hello", "hello"));
    expect(glob("h??lo", "hello"));
    expect(glob("h*o", "hello"));
    expect(glob("h*ello", "hello"));
    expect(glob("*h*o*", "hello world"));
    expect(glob("h*o*", "hello world"));
    expect(glob("*h*d", "hello world"));
    expect(glob("*h*l*w*d", "hello world"));
    expect(glob("*h?l*w*d", "hello world"));

    expect(!glob("hello", "hi"));
    expect(!glob("h?i", "hi"));
    expect(!glob("h*l", "hello"));
}

fn walk(allocator: *std.mem.Allocator, path: []const u8, pattern: []const u8) anyerror!void {
    var walker = try std.fs.walkPath(allocator, ".");
    defer walker.deinit();
    while (try walker.next()) |entry| {
        if (entry.kind == .File) {
            const file = try entry.dir.openFile(entry.basename, .{});
            defer file.close();
            const text = try file.readToEndAlloc(allocator, std.math.maxInt(usize));
            defer allocator.free(text);
            if (std.mem.indexOf(u8, text, "\x00") != null or std.mem.indexOf(u8, text, "\n") == null) {
                continue;
            }

            var lines = std.mem.tokenize(text, "\n");
            var lineno: usize = 1;
            while (lines.next()) |line| : (lineno += 1) {
                if (glob(pattern, line)) {
                    print("{s}:{d}\t{s}\n", .{ entry.path, lineno, line });
                }
            }
        }
    }
}

pub fn main() anyerror!void {
    const allocator = std.heap.page_allocator;
    const args = try std.process.argsAlloc(allocator);
    if (args.len != 2) {
        std.log.err("USAGE: {s} <pattern>", .{args[0]});
        return;
    }
    try walk(allocator, ".", args[1]);
}

File: rust/glob.rs
use std::env;
use std::io::BufRead;
use std::path::Path;
use std::{fs, io};

fn glob(pattern: &[u8], text: &[u8]) -> bool {
    let mut p: usize = 0;
    let mut t: usize = 0;
    let mut np: usize = 0;
    let mut nt: usize = 0;
    while p < pattern.len() || t < text.len() {
        if p < pattern.len() {
            match pattern[p] as char {
                '*' => {
                    np = p;
                    nt = t + 1;
                    p += 1;
                    continue;
                }
                '?' => {
                    if nt < text.len() {
                        p += 1;
                        t += 1;
                        continue;
                    }
                }
                _ => {
                    if t < text.len() && text[t] == pattern[p] {
                        p += 1;
                        t += 1;
                        continue;
                    }
                }
            }
        }
        if nt > 0 && nt <= text.len() {
            p = np;
            t = nt;
            continue;
        }
        return false;
    }
    true
}

fn walk(dir: &Path, pattern: &str) -> io::Result<()> {
    for entry in fs::read_dir(dir)? {
        let path = entry?.path();
        if path.is_dir() {
            walk(&path, &pattern)?;
            continue;
        }
        let file = fs::File::open(&path)?;
        let reader = io::BufReader::new(file);
        let mut lineno = 0;
        for line in reader.lines() {
            lineno += 1;
            let line = match line {
                Ok(s) => s,
                Err(_) => "".to_string(),
            };
            if glob(pattern.as_bytes(), line.as_bytes()) {
                println!("{}:{}\t{}", path.to_str().unwrap_or(""), lineno, line);
            }
        }
    }
    Ok(())
}

pub fn main() {
    let argv: Vec<String> = env::args().collect();
    if argv.len() != 2 {
        println!("USAGE: glob <pattern>");
        return;
    }
    match walk(Path::new("."), &argv[1]) {
        Ok(()) => (),
        Err(error) => panic!("Error: {:?}", error),
    };
}

#[cfg(test)]
mod test {
    use super::glob;
    #[test]
    fn test_glob() {
        assert!(glob(b"", b""));
        assert!(glob(b"hello", b"hello"));
        assert!(glob(b"h??lo", b"hello"));
        assert!(glob(b"h*o", b"hello"));
        assert!(glob(b"h*ello", b"hello"));
        assert!(glob(b"*h*o*", b"hello world"));
        assert!(glob(b"h*o*", b"hello world"));
        assert!(glob(b"*h*d", b"hello world"));
        assert!(glob(b"*h*l*w*d", b"hello world"));
        assert!(glob(b"*h?l*w*d", b"hello world"));

        assert!(!glob(b"hello", b"hi"));
        assert!(!glob(b"h?i", b"hi"));
        assert!(!glob(b"h*l", b"hello"));
    }
}
